feat: adding music player project (#52487)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65362bfd67d61d517deef191.md

---
id: 65362bfd67d61d517deef191
title: Step 21
challengeType: 0
dashedName: step-21
---

# --description--

> The <dfn>find()</dfn> method returns the first element of an array that satisfies the callback function passed to it. If no element is found, then it will return `undefined`.

Reviewer: ```suggestion
The <dfn>find()</dfn> method retrieves the first element within an array that fulfills the conditions specified in the provided callback function. If no element satisfies the condition, the method returns `undefined`.
```

In the example below, the `find()` method is used to find the first number greater than 25:

```js
const numbers = [10, 20, 30, 40, 50];

// Find the first number greater than 25
const foundNumber = numbers.find((number) => number > 25);
console.log(foundNumber); // Output: 30
```

Use `const` to create a variable named `song` and assign it to the `find()` method. Use pass in `song` as the parameter of the `find()`callback and check if `song.id` is strictly equal to `id`.

> This wiil scan through the `userData?.songs` array and searches for a song that matches the `id` passed into the `playSong` function.

Reviewer: ```suggestion
This will iterate through the `userData?.songs` array, searching for a song that corresponds to the `id` passed into the `playSong` function.
```

# --hints--

You should use the `find` method on `userData?.songs`.

```js
assert.match(code, /userData\?\.songs\.find\(/)
```

Your `find` method should have `song` as the parameter of its arrow function callback.

```js
assert.match(code, /userData\?\.songs\.find\(\(?song\)?/)
```

Your `find` method should use strict equality to check if `song.id` is equal to `id`.

```js

> assert.match(code, /userData\?\.songs\.find\(\(?song\)?\s*=>\s*song.id\s*===\s*id\);?/)

Author: I deliberately put `song.id` first to let campers know that's what to add first.

> assert.match(code, /userData\?\.songs\.find\(\(?song\)?\s*=>\s*song.id\s*===\s*id\);?/)

Reviewer 2: I'd say it would make sense to allow both in regex.
